### PR TITLE
Fixing warnings that result from the following clang options:

### PR DESCRIPTION
**Pull request by rmpowell77**
-Wnon-virtual-dtor
-Wshadow
-Wsign-conversion
-Wmissing-prototypes

### CODE REVIEW

File: docopt.cpp [hunk 1]
```diff
@@ -119,7 +119,7 @@ bool LeafPattern::match(PatternList& left, std::vector<std::shared_ptr<LeafPatte
 		return false;
 	}
 	
-	left.erase(left.begin()+match.first);
+	left.erase(left.begin()+static_cast<decltype(left.begin())::difference_type>(match.first));
```
Comment (jaredgrubb) [inline docopt.cpp]: Do you mind if we just use "ssize_t" in the cast directly?

File: docopt.cpp [hunk 2]
```diff
@@ -170,7 +170,7 @@ Option Option::parse(std::string const& option_description)
 	auto double_space = option_description.find("  ");
 	auto options_end = option_description.end();
 	if (double_space != std::string::npos) {
-		options_end = option_description.begin() + double_space;
+		options_end = option_description.begin() + static_cast<decltype(option_description.begin())::difference_type>(double_space);
```
Comment (jaredgrubb) [inline docopt.cpp]: ditto here

**Comment by jaredgrubb** [comment]
Great changes! I'd like to simplify the casts. Casts are ugly to begin with, but I'd rather just use "ssize_t" directly because there's no STL that would actually make "distance_type" different than that in practice, right?

**Comment by rmpowell77** [comment]
Now merge-able

**Comment by GamePad64** [comment]
There is no ssize_t in the C++ Standard. It is unportable POSIX-only feature.
Use ptrdiff_t instead.

**Comment by rmpowell77** [comment]
Using ptrdiff_t.  Good to go!

**Comment by rmpowell77** [comment]
Fixed in my latest checkin

**Comment by jaredgrubb** [comment]
Hmm... TravisCI doesnt like this. Says "ptrdiff_t" is not allowed. I think you need std:: there.

**Comment by GamePad64** [comment]
Check if <stddef.h> included for ptrdiff_t

**Comment by rmpowell77** [comment]
Ok, now travis looks happy!

**Comment by GamePad64** [comment]
Oh, the text of my message was `Check if <stddef.h> included for ptrdiff_t`. GitHub parser is stupid enough to cut `<>` off, if not backtick'ed.

**Comment by jaredgrubb** [comment]
Awesome! Thanks everyone!

**Pull request merged**